### PR TITLE
[Bugfix] Fix layerwise reload dropping params after a composed weight loader

### PR DESCRIPTION
**Pull request by hallerite**
## Purpose

Fixes a bug in the layerwise weight-reload path
(`vllm/model_executor/model_loader/reload`) that drops a trailing parameter of a
layer when an earlier parameter uses `composed_weight_loader`.

The reload path counts loaded elements per layer (`CopyCounter` inside
`get_numel_loaded`) and finalizes a layer once `load_numel >= load_numel_total`.
`composed_weight_loader` copies into its parameter **twice** — the initial load
plus an in-place post-load transform `param.copy_(fn(param))` — so `CopyCounter`
reports **2×** that parameter's size. The over-count can push `load_numel` to the
layer total before every parameter has been loaded, finalizing the layer early
and silently dropping the parameter(s) loaded afterward. The dropped parameter is
left as `materialize_meta_tensor` (`empty_strided`) memory, i.e. uninitialized
garbage.

This bites Mamba2 mixers (`MambaMixer2`, used by NemotronH and other hybrid SSM
models). The mixer's `A`, `D`, and `dt_bias` are equal-numel direct parameters,
and `A` is loaded with `composed_weight_loader` (`-exp(A_log)`). On an online
weight reload, `A` counts double, the mixer finalizes after `A` + one of the
others, and `mixer.D` (the SSD skip connection, loaded last) is dropped — leaving
NaN/inf in `D` and producing NaN logits after a weight update.

### Fix

A weight loader fills a single destination parameter, so the number of distinct
elements it loads is at most that parameter's size. Copying into the same
parameter more than once (the composed transform) is not additional loaded data.
`get_numel_loaded` now caps each loader call's counted elements at the
destination parameter's `numel()`, restoring the per-layer accounting invariant
("the total is reached only after all parameters are loaded"). One-line behavior
change in a single function; no loader or model code is touched.

## Test Plan

`tests/model_executor/model_loader/test_reload.py`:

- `test_get_numel_loaded_caps_at_param_size` — a `composed_weight_loader` reports
  `param.numel()`, not `2×`.
- `test_layerwise_reload_composed_loader_does_not_drop_params` — a minimal
  module mirroring a Mamba2 mixer (`A` composed-loaded, `D`/`dt_bias` plain,
  equal numel) goes through `record_metadata_for_reloading` →
  `initialize_layerwise_reload` → load (`D` last) → `finalize_layerwise_reload`;
  asserts all three params keep their loaded values. Materialized tensors are
  filled with NaN so a dropped param is detectable.

```
pytest tests/model_executor/model_loader/test_reload.py -k \
  "get_numel_loaded or composed_loader or alias_buffers"
```

## Test Result

- Both new tests **fail on `main`** (`D` is NaN; `get_numel_loaded` returns
  `2×numel`) and **pass with this change**.
- Existing reload unit tests (`test_get_numel_loaded`,
  `test_layerwise_reload_skips_*_alias_buffers`) continue to pass.
- `ruff check` and `ruff format --check` pass on both files.

This is not a duplicate of #40647 (which fixes the `VllmConfig` context and the
alias-buffer copy-back in the same file) — that PR does not touch the
`get_numel_loaded` element accounting that drops `mixer.D`.

---

This change was AI-assisted (Claude Code). The root cause, fix, and tests were
reviewed by the submitter, who ran the tests and linters above.

### CODE REVIEW

**Review by SumanthRH — APPROVED** [review]
Great catch!

File: vllm/model_executor/model_loader/reload/meta.py [hunk 1]
```diff
@@ -185,4 +185,18 @@ def get_numel_loaded(
...
+    # loaded elements is at most that parameter's size. Some loaders copy into
+    # the parameter more than once -- e.g. ``composed_weight_loader`` runs an
+    # in-place post-load transform (``param.copy_(fn(param))``) on top of the
+    # initial copy -- which would make CopyCounter report twice the parameter
+    # size. Over-counting inflates the layer's loaded-element total and can
+    # finalize the layer before every parameter is loaded, silently dropping
+    # the trailing parameter(s) (e.g. Mamba ``mixer.D``). Cap the count at the
+    # destination size to keep the per-layer accounting correct.
+    numel = counter.copied_numel
+    param = args.arguments.get("param")
```
Comment (kylesayrs) [inline vllm/model_executor/model_loader/reload/meta.py]:
```suggestion
    param = args.arguments.get("param", None)
```

**Review by kylesayrs — COMMENTED** [review]
Nice addition. But please check the `param` kwarg to avoid errors with unconventional loaders

**Review by kylesayrs — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
